### PR TITLE
Add a new null safety example

### PR DESCRIPTION
**Pull request by mit-mit**
TODOs:
  - [x] Await fix of https://github.com/flutter/flutter/issues/59879#issuecomment-704555165
  - [x] Await fix of https://github.com/flutter/flutter/issues/68433
  - [ ] Update Flutter SDK lower constraint in `pubspec.yaml` to the build number that has fixes to above bugs
    - [x] Update to min master channel build
    - [ ] If possible, update to min dev channel build
  - [x] Confirm that `null_safety/null_safe_app` can now be run:
    - [x] In terminal: `flutter run --enable-experiment=non-nullable`
    - [x] In VSCode: Open folder and hit F5 (it has the experiment flag in the launch config): blocked on https://github.com/flutter/flutter/issues/68627
    - [x] In Android Studio: Open folder and hit F5 (it has the experiment flag in the launch config): blocked on https://github.com/flutter/flutter/issues/68627
  - [x] Confirm that `null_safety/null_safe_app` tests now pass: `flutter test --enable-experiment=non-nullable`
  - [ ] Update https://flutter.dev/docs/null-safety to link to this sample. (@kwalrath)

### CODE REVIEW

**Comment by mit-mit** [comment]
cc @lrhn can you please review the Dart code in `null_unsafe_app/lib` and `null_safe_app/lib` ?

File: null_safety/README.md
Comment (lrhn) [inline null_safety/README.md:10]: error -> errors
it's -> its
drop comma

File: null_safety/README.md
Comment (lrhn) [inline null_safety/README.md:13]: Missing end-\`.

File: null_safety/README.md
Comment (lrhn) [inline null_safety/README.md:14]: Can we use a different word than "spurious". I had to look up that it can also mean:
> of a deceitful nature or quality

to understand what was meant. Maybe "sporadic"? Maybe something better.

File: null_safety/README.md
Comment (lrhn) [inline null_safety/README.md:40]: Maybe add "(on macOS)" and "(on Windows)" (if that's what the paths are).
(None of them are Linux, that probably uses `/home/michael/` instead).

File: null_safety/README.md
Comment (lrhn) [inline null_safety/README.md:46]: I'd use `_experiment flag_` (_experiment flag_) instead of single quotes. Not sure what single quotes mean in English, but if it's intended to mean the same as double quotes, then it's redundant with the "so-called" prefix.

File: null_safety/README.md
Comment (lrhn) [inline null_safety/README.md:55]: So, not Linux here.

File: null_safety/README.md
Comment (lrhn) [inline null_safety/README.md:62]: "runs passing" -> "passes".

File: null_safety/null_safe_app/lib/goodapp.dart
Comment (lrhn) [inline null_safety/null_safe_app/lib/goodapp.dart:37]: "null error" is vague here. It sound like a run-time error ("value is `null`") not the compile-time "variable is potentially `null`" that one could expect here.

Maybe:
// Notice how we are not getting an error saying that `tempWidgets` 
// is uninitialized or potentially `null` despite it being declared ....

File: null_safety/null_safe_app/lib/goodapp.dart
Comment (lrhn) [inline null_safety/null_safe_app/lib/goodapp.dart:39]: definete -> definite

Consider newline after `:`. The `We` feels dangling.

**Review by lrhn — APPROVED** [review]

File: null_safety/null_safe_app/lib/goodapp.dart
Comment (lrhn) [reply null_safety/null_safe_app/lib/goodapp.dart:39]: Also: assigment -> assignment

File: null_safety/README.md
Comment (mit-mit) [reply null_safety/README.md:55]: I think it's close enough on Linux (plus Linux users are tech savvy) so I think they can grok the intent

**Comment by mit-mit** [comment]
@johnpryan I'm heading out on vacation in a few hours. Any chance you could take this PR to completion? I listed some TODO steps in the top comment. And I enabled edits by maintainers, so you could be able to commit into the PR.

**Comment by johnpryan** [comment]
@mit-mit happy to help. cc: @RedBrogdon should we move this into the experimental/ directory?

**Comment by RedBrogdon** [comment]
A) This is awesome, and thanks, @mit-mit, for contributing something for null safety to the repo.
B) Given that there's an outstanding bug blocking this from landing and that there are some meta-issues that would need to be addressed (how does/can CI work for this sample, for example?), I'd really prefer to wait to merge this until you're back from vacation and we can talk over a couple things. If there's a pressing need to merge it now, though, folks are free to push back.

File: null_safety/README.md
Comment (kwalrath) [inline null_safety/README.md:26]:
```suggestion
    `lib/badapp.dart`: This is the same source code as in the unsafe app. Try
    uncommenting the two commented lines of null-unsafe code, and notice how the
    Dart analyzer with null safety immediately flags the lines as unsafe.

    `lib/goodapp.dart`: This a safe version of the unsafe app, with the null
    errors fixed. Read through the comments to see illustrations of how to resolve
    null errors with null safety.
```

File: null_safety/README.md
Comment (kwalrath) [inline null_safety/README.md:4]:
```suggestion
safety. For additional information about this feature, see
```

File: null_safety/README.md
Comment (kwalrath) [inline null_safety/README.md:41]:
```suggestion
     macOS/Linux) or `C:\Users\michael\dev\preview\flutter\` (on Windows).
```

File: null_safety/README.md
Comment (kwalrath) [inline null_safety/README.md:36]:
```suggestion
  1. Download the latest null safety preview build from the **dev channel** of
```

**Review by kwalrath — COMMENTED** [review]
Drive-by tweaks to the README.

**Comment by mit-mit** [comment]
@RedBrogdon happy to chat! What is the CI concern? This is passing CI for PRs. Do we use a different job past merge?

**Comment by mit-mit** [comment]
> should we move this into the experimental/ directory?

Andrew and I discussed and agreed it should. I has been moved in this PR.

**Comment by johnpryan** [comment]
I filed https://github.com/flutter/flutter/issues/68634 - I'm seeing an exception in the framework when I run the app in IntelliJ

File: experimental/null_safety/README.md
Comment (RedBrogdon) [inline experimental/null_safety/README.md:20]: If the main educational point is to compare the two approaches, WDYT about removing null_unsafe_app entirely, and focusing on null_safe_app and the difference between its two primary Dart files?

File: experimental/null_safety/README.md
Comment (RedBrogdon) [inline experimental/null_safety/README.md:36]: We have some instructions in the README for the experimental directory that instruct people to switch to the master channel for the samples in this folder:

https://github.com/flutter/samples/tree/master/experimental

If it's better for people to use the dev channel, that's fine too, but consider just telling folks to `flutter channel dev`. It can be safely assumed that they've got the SDK installed.

File: experimental/null_safety/README.md
Comment (RedBrogdon) [inline experimental/null_safety/README.md:52]: Is `calculate_lix` left over from an earlier version?

File: experimental/null_safety/README.md
Comment (RedBrogdon) [inline experimental/null_safety/README.md:93]: Is that version number stale?

File: experimental/null_safety/null_safe_app/lib/badapp.dart
Comment (RedBrogdon) [inline experimental/null_safety/null_safe_app/lib/badapp.dart:1]: All the Dart files need copyright headers. Should be able to use the same one as you'll find in the other apps.

File: experimental/null_safety/null_safe_app/lib/goodapp.dart
Comment (RedBrogdon) [inline experimental/null_safety/null_safe_app/lib/goodapp.dart:13]: Excellent use of emoticon. :)

File: experimental/null_safety/null_safe_app/lib/goodapp.dart
Comment (RedBrogdon) [inline experimental/null_safety/null_safe_app/lib/goodapp.dart:15]: Nit: In general, it's best to avoid first-person pronouns (I, we) in code comments. My usual cheat is to say "The app" or "this method" instead, as in "This method checked for null..."

File: experimental/null_safety/null_unsafe_app/test/widget_test.dart
Comment (RedBrogdon) [inline experimental/null_safety/null_unsafe_app/test/widget_test.dart:1]: I don't know that this would pass. You could either cut it entirely or copy over the one from the other app.

**Review by RedBrogdon — CHANGES_REQUESTED** [review]

File: experimental/null_safety/README.md
Comment (mit-mit) [reply experimental/null_safety/README.md:20]: I thought about it, but I'm not sure that motivates the problem. Some might not understand how they are prone to null errors today. `null_unsafe_app` is meant to show that in a very viceral way, with an app that works some of the time and fails the rest of it.

In `null_safe_app`, perhaps my naming `goodapp.dart` and `badapp.dart` is misleading. They both have null safety enabled, it's just that one doesn't have the errors fixed yet. Perhaps they should be named `app_with_null_errors.dart` and `app.dart`?

File: experimental/null_safety/null_safe_app/lib/goodapp.dart
Comment (mit-mit) [reply experimental/null_safety/null_safe_app/lib/goodapp.dart:13]: :-)

File: experimental/null_safety/README.md
Comment (mit-mit) [reply experimental/null_safety/README.md:20]: We could also rename the top-level folders to `with_null_safety` (`null_safe_app` currently), and `without_null_safety` (`null_unsafe_app` currently)

File: experimental/null_safety/README.md
Comment (mit-mit) [reply experimental/null_safety/README.md:36]: Done

File: experimental/null_safety/README.md
Comment (mit-mit) [reply experimental/null_safety/README.md:52]: Good catch, fixed

File: experimental/null_safety/null_safe_app/lib/badapp.dart
Comment (mit-mit) [reply experimental/null_safety/null_safe_app/lib/badapp.dart:1]: Yes, of course

File: experimental/null_safety/null_unsafe_app/test/widget_test.dart
Comment (mit-mit) [reply experimental/null_safety/null_unsafe_app/test/widget_test.dart:1]: Removed

File: experimental/null_safety/README.md
Comment (mit-mit) [reply experimental/null_safety/README.md:93]: Yes, and so were some instructions, updated

File: experimental/null_safety/README.md
Comment (RedBrogdon) [reply experimental/null_safety/README.md:20]: Let's discuss this offline at some point in the coming week. Right now, you've got a clear description of things in the README, and the actual code is tight enough that people sophisticated enough to be looking at NNBD right now will have no trouble understanding what's going on.

**Review by RedBrogdon — APPROVED** [review]

**Pull request merged**
